File: src/p132/Solution.java
import java.util.Arrays;

public class Solution {
	public int minCut(String s) {
		int n = s.length();
		boolean[][] palin = new boolean[n][];
		for (int i = 0; i < n; i++) {
			palin[i] = new boolean[n];
			palin[i][i] = true;
		}
		for (int l = 2; l <= n; l++) {
			for (int i = 0; i < n; i++) {
				int e = Math.min(n - 1, i + l - 1);
				palin[i][e] = s.charAt(i) == s.charAt(e) && (i + 1 >= e - 1 || palin[i + 1][e - 1]);
			}
		}

		int[] d = new int[n + 1];
		Arrays.fill(d, -1);
		d[0] = 0;

		boolean[] g = new boolean[n+1];
		for (int i = 0; i < n; i++) {
			int vertex = -1;
			int dist = Integer.MAX_VALUE;
			for (int j = 0; j <= n; j++) {
				if (!g[j] && d[j] != -1 && d[j] < dist) {
					vertex = j;
					dist = d[j];
				}
			}
			g[vertex] = true;
			if(vertex >= n)
				continue;
			for (int j = 1; j <= n; j++) {
				if (palin[vertex][j-1] && (d[j] == -1 || d[j] > d[vertex] + 1))
					d[j] = d[vertex] + 1;
			}
		}
		return d[n] - 1;
	}

	public static void main(String[] args) {
		System.out.println(new Solution().minCut("aabbccc"));
	}
}

File: src/p132/solution.cpp
class Solution {
public:
    int minCut(string s) {
        int n = s.length();
        vector<vector<bool>> palin(n, vector<bool>(n, false));
        for (int k = 1 ; k <= n ; k++) {
            for (int i = 0 ; i + k - 1 < n ; i++) {
                int j = i + k - 1;
                if (j == i || s[i] == s[j] && (i + 1 == j || palin[i+1][j-1]))
                    palin[i][j] = true;
            }
        }
        vector<int> dp(n, -1);
        dp[0] = 0;
        for (int i = 0; i < n ; i++) {
            for (int j = 0; j <= i; j++) {
                if (palin[j][i]) {
                    int cut = (j == 0 ? 0 : dp[j-1] + 1);
                    dp[i] = (dp[i] == -1 ? cut : min(dp[i], cut)); 
                }
            }
        }
        return dp[n-1];
    }
};
